wgpu-core/src/device/queue.rs

```diff
@@ -305,30 +305,141 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
+        let (staging_buffer, staging_buffer_ptr) = device.prepare_staging_buffer(data_size)?;
+
+        profiling::scope!("copy");
         unsafe {
-            profiling::scope!("copy");
-            stage.write(&device.raw, 0, data)
+            ptr::copy_nonoverlapping(data.as_ptr(), staging_buffer_ptr, data.len());
+            staging_buffer.flush(&device.raw)?;
+        };
+
+        let mut trackers = device.trackers.lock();
+        let (dst, transition) = trackers
+            .buffers
+            .set_single(&*buffer_guard, buffer_id, hal::BufferUses::COPY_DST)
+            .ok_or(TransferError::InvalidBuffer(buffer_id))?;
+        let dst_raw = dst
+            .raw
+            .as_ref()
+            .ok_or(TransferError::InvalidBuffer(buffer_id))?;
+        if !dst.usage.contains(wgt::BufferUsages::COPY_DST) {
+            return Err(TransferError::MissingCopyDstUsageFlag(Some(buffer_id), None).into());
         }
-        .map_err(DeviceError::from)?;
+        dst.life_guard.use_at(device.active_submission_index + 1);
+
+        if data_size % wgt::COPY_BUFFER_ALIGNMENT != 0 {
+            return Err(TransferError::UnalignedCopySize(data_size).into());
+        }
+        if buffer_offset % wgt::COPY_BUFFER_ALIGNMENT != 0 {
+            return Err(TransferError::UnalignedBufferOffset(buffer_offset).into());
+        }
+        if buffer_offset + data_size > dst.size {
+            return Err(TransferError::BufferOverrun {
+                start_offset: buffer_offset,
+                end_offset: buffer_offset + data_size,
+                buffer_size: dst.size,
+                side: CopySide::Destination,
+            }
+            .into());
+        }
+
+        let region = wgt::BufferSize::new(data_size).map(|size| hal::BufferCopy {
+            src_offset: 0,
+            dst_offset: buffer_offset,
+            size,
+        });
+        let barriers = iter::once(hal::BufferBarrier {
+            buffer: &staging_buffer.raw,
+            usage: hal::BufferUses::MAP_WRITE..hal::BufferUses::COPY_SRC,
+        })
+        .chain(transition.map(|pending| pending.into_hal(dst)));
+        let encoder = device.pending_writes.activate();
+        unsafe {
+            encoder.transition_buffers(barriers);
+            encoder.copy_buffer_to_buffer(&staging_buffer.raw, dst_raw, region.into_iter());
+        }
+
+        device.pending_writes.consume(staging_buffer);
+        device.pending_writes.dst_buffers.insert(buffer_id);
+
+        // Ensure the overwritten bytes are marked as initialized so they don't need to be nulled prior to mapping or binding.
+        {
+            drop(buffer_guard);
+            let (mut buffer_guard, _) = hub.buffers.write(&mut token);
+
+            let dst = buffer_guard.get_mut(buffer_id).unwrap();
+            dst.initialization_status
+                .drain(buffer_offset..(buffer_offset + data_size));
+        }
+
+        Ok(())
+    }
+
+    pub fn queue_create_staging_buffer<A: HalApi>(
```

Comment:
My only real concern is if we can get these two methods of writing buffers to share mode code.